Add ABI language to the standard

jjhursey opened the pull request:
* Fixes #364

jjhursey commented:
Leaving this in draft form while we discuss #364 - this is an attempt to start sketching the language.

jjhursey commented on Chap_Introduction.tex:
```diff
@@ -233,3 +234,39 @@ \subsection{PMIx Roles}
+\item Attributes that are required to be implemented and their string representation
+\end{compactalphaenum}
+
+When adding a field to an existing data structure that has padding then that padding must be adjusted so as to preserve the stability of the \ac{ABI}.
+If a data structure must be changed in a way that breaks the \ac{ABI} the authors should consider creating a new data structure if possible.
+Additionally, the authors must take care to update or duplicate existing PMIx interfaces that are impacted by the change.
```
Remove. No padding in the current standard. May add in the future if padding in a structure is added.

jjhursey commented:
Please use emoji reactions ON THIS COMMENT to indicate your position on this proposal.

* You do not need to vote on every proposal
* If you have no opinion, don't vote - that is also useful data
* If you've already commented on this issue, please still vote so
      we know your current thoughts
* Not all proposals solve exactly the same problem, so we may end
      up accepting proposals that appear to have some overlap
This is not a binding majority-rule vote, but it will be a very
significant input into the corresponding ASC decision.

Here are the meanings for the emojis:

* Hooray or Rocket: I support this so strongly that I
      want to be an advocate for it
* Heart: I think this is an ideal solution
* Thumbs up: I'd be happy with this solution
* Confused: I'd rather we not do this, but I can tolerate it
* Thumbs down: I'd be actively unhappy, and may even consider
      other technologies instead
If you want to explain in more detail, feel free to add another
comment, but please also vote on this comment.

raffenet commented on Chap_API_Struct.tex:
```diff
@@ -969,6 +969,11 @@ \subsection{Value Structure}
 }
 
+The size of the \refstruct{pmix_value_t} must remain constant on a given platform regardless of the compiler.
+The \refstruct{pmix_value_t} is \code{32} bytes with \code{24} bytes reserved for the \code{data} union member.
+An implementation may add a padding member to the structure to achieve the required 32-byte size.
```
Are we saying it should be 32 bytes regardless of platform or just on x86-64 (or ppc64le 😄 )? FWIW, if I build a 32-bit x86 executable with the current ABI headers,`pmix_value_t` is 16 bytes and the data portion is 12.

rhc54 replied:
Good point - we definitely don't want to force 32-bit builds to waste that space.

jjhursey replied:
Good point. Do you all have a suggestion on a re-wording?

rhc54 replied:
Actually, let's take a step back. It occurs to me that we may be over complicating things here. Let's take a look at what we mean by "ABI".

I think we all agree that we should take the "build" meaning for ABI, so let's take that as a starting point. This means that a program is assured that they can access a library of the same ABI they built against, with some assurance that symbols will resolve, constants will match in value, and structs will have the same size/ordering of fields.

There is nothing in that statement to say a 64-bit program could use a 32-bit library, or vice versa. Likewise, there is no statement that indicates you can move your app across architecture without change - which is good because that is pretty much impossible anyway. If you move from x86-64 to ARM, you are going to have to recompile.

So any statements about padding should be restricted by those constraints. It should be adequate to state that the padding should align values with the native (64 or 32-bit) addressing scheme.

However, there is also nothing that proposes that a client app can interact with a server based on a different implementation. In the case of a shared memory subsystem, that would be highly unlikely to work anyway as the details of internal implementation are probably different. For non-shared memory interactions, there is no standard wire protocol - and even if there were, it wouldn't affect the ABI as both sides instantiate the objects passed over the wire themselves.

So do we even need padding? I'm thinking we don't as the only reason to pad was the shared memory subsystem, and that actually is an implementation-internal subsystem - and not an ABI. In other words, the ABI should not apply to cross-implementation interactions - it should be restricted to the interface between a program and an implementation's library.

jjhursey replied:
The primary use case that we have for the ABI is to be able to build against the PMIx headers, then link with various PMIx implementations. This would allow a tool to build without linking to a specific PMIx implementation then use `dlopen/dlsym` to pick up the library symbols at runtime.

On a given architecture+compiler the sizeof the `pmix_value_t` should not change if the structure stays the same as it is defined in the standard. If additional fields are added to the union by an implementation beyond what is defined in the standard then it should not exceed the sizeof the data union as it exists in the standard.

Maybe that last sentence is all that we need? That will ensure that a tool built against ABI vX.Y will work with an implementation that supports ABI vX.Y even if that implementation has extra elements in the data union - since that union has not grown. So any arrays of structures that the tool has created statically are still at the correct size/offset.

I'd like to discuss this today in the call to reach a resolution.

raffenet replied:
> Maybe that last sentence is all that we need? That will ensure that a tool built against ABI vX.Y will work with an implementation that supports ABI vX.Y even if that implementation has extra elements in the data union - since that union has not grown. So any arrays of structures that the tool has created statically are still at the correct size/offset.

IMO, yes. That should be all we need. As long as the size (and offset) of the elements in the structure are consistent, then no pad should be needed.

jjhursey replied:
Note from April 14, 2022 meeting:
> The PMIx ABI is platform-specific. The headers are defined, but the realization of the ABI in binary form is platform-specific. Therefore care must be taken with data structures to not change the position of members, size of the structure or members in the structure, and the offset of the members in the structure for that platform.

jjhursey replied:
I just pushed this change as a new commit.

jjhursey commented:
I pushed a new pair of commits addressing the comments from the discussion so far. Mirrors over the versioning text from the Governance document [here](https://github.com/pmix/governance/pull/34/files#r851349685)

abouteiller commented on Chap_Introduction.tex:
```diff
@@ -232,4 +234,81 @@ \subsection{PMIx Roles}
+\item Attributes that are required to be implemented and their string representation
+\end{compactitemize}
+
+If a data structure must be changed in a way that breaks the \ac{ABI} the authors should consider creating a new data structure if possible.
```
Changing the ABI of data structures (e.g., removing/reordering existing fields) will break backward compatibility. We should add a short rationale as to why this is **very** undesirable, and not only one should *consider*, but should avoid this type of changes at all cost.

jjhursey replied:
```suggestion
If a data structure must be changed in a way that breaks the \ac{ABI} the authors are strongly advised to create a new data structure instead of breaking the \ac{ABI} and backward compatibility.
```

Let's add a clear rationale to add to this section to capture the middleware constraints dealing with cross and backwards compatibility issues.

abouteiller replied:
```diff
+Changing data structures (in particular changes that modify the field alignment) and function signatures
+breaks the \ac{ABI}. The \ac{ASC} strongly discourages such changes. It is preferred that the authors 
+duplicate existing PMIx interfaces that are impacted by the change, 
+and potentially initiate the deprecation procedure for the duplicated interfaces.
+
+\rationalestart
+Breaking the \ac{ABI} of data structures and function signatures while using the same
+name prevents the implementation of backward compatible and cross-version compatible
+\ac{PMIx} libraries. 
+\rationaleend
```

abouteiller replied:
I like the new text but we may need some more plenary time to discuss the implications.

jjhursey replied:
```suggestion
The \ac{ASC} prohibits data structure and function signature changes that break the \ac{ABI}
(e.g., modifications to data structure member alignment and/or size).
It is preferred that the authors create a new data structure or function and
duplicate all impacted PMIx interfaces.

\rationalestart
Breaking the \ac{ABI} of data structures and function signatures while using the same
name prevents a PMIx implementation from supporting backward compatible and
cross-version compatible \ac{PMIx} libraries. 
\rationaleend
```

abouteiller commented on Chap_Introduction.tex:
```diff
@@ -232,4 +234,81 @@ \subsection{PMIx Roles}
+If a data structure must be changed in a way that breaks the \ac{ABI} the authors should consider creating a new data structure if possible.
+Additionally, the authors must take care to update or duplicate existing PMIx interfaces that are impacted by the change.
+
+Attributes that are required by the PMIx standard are a required part of the \ac{ABI}.
```
Ralph comment: Attributes are non-linkable string objects (i.e., they are defines, they don't have an address as a variable), hence they are not an ABI element.

abouteiller replied:
```Although the values of attributes and constants are not part of the \ac{ABI}, they are defined as standardized members of the \ac{API}, hence their value can be assumed based on the \ac{ABI} version.```

jjhursey replied:
Please take time before Wednesday afternoon to suggest any revisions to this text.

This text will replace the full paragraph here (lines 266-271).

jjhursey replied:
I captured this change in the https://github.com/pmix/pmix-standard/pull/375/files#r870587051
Let's follow up on that comment.

jjhursey commented on Chap_Introduction.tex:
```diff
@@ -232,4 +234,81 @@ \subsection{PMIx Roles}
+\item \texttt{MAJOR} incremented when the ABI changed in a backward-incompatible manner.
+\item \texttt{MINOR} incremented when functionality is added to the ABI in a backward-compatible manner.
+\end{compactitemize}
+
```
Add a comment along the lines of:
```
The PMIx Standard ABI versioning follows the Semantic versioning procedure. However, the PMIx Implementation may use a different version numbering for their library objects than the PMix Standard ABI.
```

jjhursey replied:
Add the comment:
```suggestion
The PMIx Standard ABI version numbers largely follow the Semantic Versioning 2.0.0 specification\footnote{Semantic Versioning \url{https://semver.org/spec/v2.0.0.html}}.
However, a PMIx implementation may use a different version numbering technique for the objects associated with that implementation.
```

jjhursey commented on Chap_Introduction.tex:
```diff
@@ -232,4 +234,81 @@ \subsection{PMIx Roles}
+If a data structure must be changed in a way that breaks the \ac{ABI} the authors should consider creating a new data structure if possible.
+Additionally, the authors must take care to update or duplicate existing PMIx interfaces that are impacted by the change.
+
+Attributes that are required by the PMIx standard are a required part of the \ac{ABI}.
+Attributes that are not required by the PMIx standard may be included in the \ac{ABI} while others may not.
+Attributes that are part of the \ac{ABI} are denoted as such in their description.
+PMIx attributes that are not required by the PMIx standard have a well-defined behavior (see Section~\ref{chap:intro:attributes}) when the string representation is not recognized by the implementation.
+The caller of the interface already expects to respond to the scenario in which the implementation does not support that optional attribute.
+Excluding these optional attributes from the ABI allows an implementation to provide additional attributes that are non-standard.
```
```suggestion
The value associated with attributes and constants are part of an \ac{API} and not part of a \refterm{Linker ABI}.
The PMIx Standard defines a \refterm{Build ABI} where the value associated with PMIx attributes and constants can be assumed based on the \ac{ABI} version.
```

jjhursey commented:
PMIx ASC 2Q 2022
 * Passed the first vote: 7 yes / 3 no / 1 abstain
 * Passed the revision exception vote: 9 yes / 3 no / 1 abstain
 * This (with revision exception) is eligible for second vote in the 3Q 2022 ASC quarterly meeting

jjhursey commented on Chap_Introduction.tex:
```diff
@@ -232,4 +234,81 @@ \subsection{PMIx Roles}
+\end{compactitemize}
+
+If a data structure must be changed in a way that breaks the \ac{ABI} the authors should consider creating a new data structure if possible.
+Additionally, the authors must take care to update or duplicate existing PMIx interfaces that are impacted by the change.
```
```suggestion
```

jjhursey commented:
Revision exception changes have been committed to this PR as a separate commit.

jjhursey commented:
PMIx ASC 3Q 2022
 * Passed the second vote: 8 yes / 0 no / 0 abstain
 * This will be included in the next major release.

jjhursey commented:
Rebased branch. No changes made.
